Sources/FoundationEssentials/Subprocess/Subprocess+AsyncBytes.swift

```diff
@@ -0,0 +1,110 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+import SystemPackage
+import Dispatch
+
+extension Subprocess {
+    public struct AsyncBytes: AsyncSequence, Sendable {
+        @inline(__always) static var bufferSize: Int {
+            16384
+        }
+        public typealias Element = UInt8
+
+        @_nonSendable
+        public struct Iterator: AsyncIteratorProtocol {
+            public typealias Element = UInt8
+
+            private let fileDescriptor: FileDescriptor
+            private var buffer: [UInt8]
+            private var currentPosition: Int
+            private var finished: Bool
+
+            internal init(fileDescriptor: FileDescriptor) {
+                self.fileDescriptor = fileDescriptor
+                self.buffer = []
+                self.currentPosition = 0
+                self.finished = false
+            }
+
+            private mutating func reloadBufferAndNext() async throws -> UInt8? {
+                if self.finished {
+                    return nil
+                }
+                try Task.checkCancellation()
+                do {
+                    self.buffer = try await self.read(
+                        from: self.fileDescriptor,
+                        maxLength: AsyncBytes.bufferSize)
+                    self.currentPosition = 0
+                    if self.buffer.count < AsyncBytes.bufferSize {
+                        self.finished = true
+                    }
+                } catch {
+                    self.finished = true
+                    throw error
+                }
+                return try await self.next()
+            }
+
+            public mutating func next() async throws -> UInt8? {
+                if currentPosition < buffer.count {
+                    let value = buffer[currentPosition]
+                    self.currentPosition += 1
+                    return value
+                }
+                return try await self.reloadBufferAndNext()
+            }
+
+            private func read(from fileDescriptor: FileDescriptor, maxLength: Int) async throws -> [UInt8] {
+                return try await withCheckedThrowingContinuation { continuation in
+                    DispatchIO.read(
```

Comment:
I personally don't think we should use the `Dispatch` APIs here for our async reading/writing of pipes. We can model this with task executors and can therefore keep all the code in concurrency land and properly handle cancellation.

Reply:
This NIO PR https://github.com/apple/swift-nio/pull/2648 shows how a non blocking read/write can be done via kqueue. We just need to extend that to epoll and it will work on Linux as well. While I agree this is an implementation detail we **should not** release this without handling cancellation; otherwise, this becomes a big foot gun when using the new APIs.